Add interpretable models

csinva opened the pull request:
*Issue #, if available:*

*Description of changes:* Added interpretable models.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Innixma commented on tabular/src/autogluon/tabular/models/imodels/imodels_models.py:
```diff
@@ -0,0 +1,169 @@
+
+        if is_train:
+            categorical_featnames = self._get_types_of_features(X)
+            self._feature_generator = OheFeaturesGenerator(cats_cols=categorical_featnames['categorical']) # LabelEncoderFeatureGenerator(verbosity=0)
+            self._feature_generator.fit(X=X)
```
Why use `OheFeaturesGenerator` specifically? I probably wouldn't recommend it as this is an old class that should probably be reworked. To my knowledge this will not function the way you intend, as it will drop any non-categorical features. For now, add a simple sklearn OHE logic:

```
from sklearn.preprocessing import OneHotEncoder
```

```
X = super()._preprocess(X, **kwargs)
if is_train:
    self._categorical_featnames = self._feature_metadata.get_features(valid_raw_types=['category'])
    self._other_featnames = self._feature_metadata.get_features(invalid_raw_types=['category'])
    if self._categorical_featnames:
        self._ohe = OneHotEncoder(dtype=np.uint8, handle_unknown='ignore')
        self._ohe.fit(X=X[self._categorical_featnames])
        self._ohe_columns = self._ohe.get_feature_names_out()

if self._ohe is not None:
    X_index = X.index
    X_ohe = self._ohe.transform(X[self._categorical_featnames])
    X_ohe = pd.DataFrame.sparse.from_spmatrix(X_ohe, columns=self._ohe_columns, index=X_index)
    if self._other_featnames:
        X = pd.concat([X[self._other_featnames], X_ohe], axis=1)
    else:
        X = X_ohe

return X.fillna(0)
```

This should be significantly faster (4x+).

csinva replied:
Switched over to this!

Innixma commented on tabular/setup.py:
```diff
@@ -50,11 +50,14 @@
 all_requires = []
 # TODO: Consider adding 'skex' to 'all'
-for extra_package in ['lightgbm', 'catboost', 'xgboost', 'fastai']:
+for extra_package in ['lightgbm', 'catboost', 'xgboost', 'fastai', 'imodels']:
```
Since `imodels` will probably not be a by-default installed dependency, we will need to change this logic prior to merging PR, because `autogluon` installs `autogluon.tabular[all]` during `pip install autogluon`. We may want to change it to install `autogluon.tabular[core]` (with core being the current `all`) and add `imodels` in `all` or require `imodels` to be installed separately from `all`.

Innixma replied:
CI can still use `imodels` so long as we install `imodels` in the CI run.

csinva replied:
switched over!

Innixma commented on core/src/autogluon/core/utils/try_import.py:
```diff
@@ -171,3 +173,11 @@ def try_import_rapids_cuml():
+    except ImportError:
+        raise ImportError(
+            "Unable to import dependency imodels. "
+            "A quick tip is to install via `pip install imodels`. ")
```
nit: add newline (PEP8)

csinva replied:
thanks, fixed!

Innixma reviewed (commented):
Great stuff! Added some initial review comments.

Note that this isn't a full review, as I haven't looked at all the files that have been changed.

stonehead8111 commented on tabular/src/autogluon/tabular/models/imodels/imodels_models.py:
```diff
@@ -0,0 +1,161 @@
+        super().__init__(**kwargs)
+
+    def get_model(self):
+        from imodels import RuleFitRegressor, RuleFitClassifier
```
-author stonehead8111

szha commented:
Job PR-1355-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/9/index.html

csinva commented on docs/tutorials/tabular_prediction/tabular-interpretability.md:
```diff
@@ -0,0 +1,42 @@
+
+The rule-based models take slightly different forms (see below), but all try to optimize predictive performance using as few rules as possible. See [imodels package](https://github.com/csinva/imodels) for more details. 
+
+![](https://raw.githubusercontent.com/csinva/imodels/master/docs/img/model_table.png)
```
good idea - cropped!

csinva commented on tabular/src/autogluon/tabular/configs/hyperparameter_configs.py:
```diff
@@ -84,6 +84,14 @@
     },
+
+    interpretable={
+        'RULEFIT': [{'max_rules': 7}, {'max_rules': 15}, {'max_rules': 30}],
```
Just switched it in the printing (and throughout the code) to be referred to as "OptimalRuleList"

csinva commented on tabular/src/autogluon/tabular/models/imodels/imodels_models.py:
```diff
@@ -0,0 +1,169 @@
+            self._feature_generator.fit(X=X)
+        if self._feature_generator is not None:
+            X = X.copy()
+            X = X.fillna(0)
```
thank for the catch, fixed!

csinva commented on tabular/src/autogluon/tabular/models/imodels/imodels_models.py:
```diff
@@ -0,0 +1,169 @@
+        X = super()._preprocess(X, **kwargs)
+
+        if is_train:
+            categorical_featnames = self._get_types_of_features(X)
```
updated! (along with the following line).

csinva commented:
Hey @jwmueller and @Innixma , I've addressed the comments y'all made in the code (the tutorial and [this notebook](https://github.com/csinva/imodels/blob/master/notebooks/autogluon_imodels_example.ipynb) show examples).

I'm still going to work on making the printing of the imodels prettier and more uniform, but that'll happen over in the imodels repo so shouldn't affect any of this code.

Let me know what you think!

Cheers,
Chandan

szha commented:
Job PR-1355-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/13/index.html

szha commented:
Job PR-1355-14 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/14/index.html

szha commented:
Job PR-1355-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/15/index.html

szha commented:
Job PR-1355-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/16/index.html

Innixma commented on docs/tutorials/tabular_prediction/tabular-interpretability.md:
```diff
@@ -0,0 +1,45 @@
+
+![](https://raw.githubusercontent.com/csinva/imodels/master/docs/img/model_table_rules.png)
+
+In addition to the usual functions in `TabularPredictor`, this predictor fitted with interpretable models has some additional functionality. For example, we can now inspect the complexity of the fitted models (i.e. how many rules they contain).
```
Add a call to `predictor.leaderboard()` to show the models and their quality.

csinva replied:
fixed!

Innixma commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2947,6 +2957,74 @@ def _sanitize_stack_args(self, num_bag_folds, num_bag_sets, num_stack_levels, ti
+    def interpretable_models_summary(self, verbosity=0):
+        '''Summary of fitted interpretable models along with their corresponding complexities
+        '''
+        d = self.fit_summary(verbosity=verbosity)
```
Its a bit strange that this is very similar to leaderboard but has different names for the same columns in a different order. Why not just call predictor.leaderboard() internally and add a couple new columns to it? In general, I'd recommend using leaderboard rather than fit_summary.

Innixma commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2947,6 +2957,74 @@ def _sanitize_stack_args(self, num_bag_folds, num_bag_sets, num_stack_levels, ti
+            model_name = summaries.index.values[i]
+            agmodel = self._learner.load_trainer().load_model(model_name)
+            imodel = agmodel.model
+            if not isinstance(agmodel, IModelsModel):
```
use `predictor.info()` to get the model class, no need to load every model from disk to get this info.

csinva replied:
switched to `predictor.info()`

Innixma commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2947,6 +2957,74 @@ def _sanitize_stack_args(self, num_bag_folds, num_bag_sets, num_stack_levels, ti
+            if not isinstance(agmodel, IModelsModel):
+                complexities.append(np.nan)
+            else:
+                complexities.append(imodel.complexity_)
```
Instead, store this information in `predictor.info()` by specifying in IModelsModel:

```
    def get_info(self):
        info = super().get_info()
        info['complexity'] = self.model.complexity_
        return info
```

This way you can completely avoid loading any models.

csinva replied:
added!

Innixma reviewed (commented):
Great stuff! I've added a full review.

csinva commented on tabular/src/autogluon/tabular/models/imodels/imodels_models.py:
```diff
@@ -0,0 +1,153 @@
+
+class RuleFitModel(IModelsModel):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
```
fixed!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -21,10 +22,13 @@
 from autogluon.core.utils.savers import save_pkl, save_str
 from autogluon.core.utils.utils import setup_outputdir, default_holdout_frac, get_approximate_df_mem_usage
 
+
```
fixed!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -21,10 +22,13 @@
 from autogluon.core.utils import plot_performance_vs_trials, plot_summary_of_models, plot_tabular_models
 from autogluon.core.utils.decorators import apply_presets
+from autogluon.tabular.models import IModelsModel
+
```
fixed!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -977,6 +985,7 @@ def fit_extra(self, hyperparameters, time_limit=None, base_model_names=None, **k
         core_kwargs = {'ag_args': ag_args, 'ag_args_ensemble': ag_args_ensemble, 'ag_args_fit': ag_args_fit,
                        'excluded_model_types': excluded_model_types}
 
+
```
fixed!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -1734,6 +1743,7 @@ def transform_labels(self, labels, inverse=False, proba=False):
                 labels_transformed = self._learner.label_cleaner.transform(y=labels)
         return labels_transformed
 
+
```
fixed!

csinva commented on docs/tutorials/tabular_prediction/tabular-interpretability.md:
```diff
@@ -0,0 +1,45 @@
+
+In this tutorial, we will explain how to automatically use interpretable models powered by integration with [🔍 the imodels package](https://github.com/csinva/imodels). This allows for automatically learning models based on rules which are extremely concise and can be useful for (1) understanding data or (2) building a transparent predictive model.
+
+Being by loading in data to predict. Note: interpretable rule-based modeling is currently only supported for binary classification.
```
fixed!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2178,6 +2188,7 @@ def get_oof_pred_proba(self, model: str = None, transformed=False, as_multiclass
                                    f'oof_pred = predictor.get_oof_pred(train_data=train_data)\n')
                 else:
                     missing_idx = list(train_data.index.difference(y_pred_proba_oof_transformed.index))
+
```
fixed

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2860,13 +2871,11 @@ def _validate_fit_data(self, train_data, tuning_data=None, unlabeled_data=None):
             if not isinstance(tuning_data, pd.DataFrame):
                 raise AssertionError(
                     f'tuning_data is required to be a pandas DataFrame, but was instead: {type(tuning_data)}')
+
```
fixed

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2881,6 +2890,7 @@ def _validate_fit_data(self, train_data, tuning_data=None, unlabeled_data=None):
             if not isinstance(unlabeled_data, pd.DataFrame):
                 raise AssertionError(
                     f'unlabeled_data is required to be a pandas DataFrame, but was instead: {type(unlabeled_data)}')
+
```
fixed

csinva commented on .gitignore:
```diff
@@ -10,6 +10,7 @@
 *.json
 *.ag
 *.xml
+*Icon*
```
removed

csinva commented on docs/tutorials/tabular_prediction/index.rst:
```diff
@@ -67,6 +73,7 @@ For standard datasets that are represented as tables (stored as CSV file, parque
    tabular-quickstart
    tabular-indepth
    tabular-kaggle
+   tabular-interpretability
```
moved

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -3003,3 +3081,4 @@ def from_learner(cls, learner: AbstractLearner):
         predictor = cls(label=learner.label, path=learner.path)
         predictor._set_post_fit_vars(learner=learner)
         return predictor
+
```
fixed

szha commented:
Job PR-1355-17 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/17/index.html

szha commented:
Job PR-1355-18 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/18/index.html

szha commented:
Job PR-1355-19 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/19/index.html

szha commented:
Job PR-1355-20 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/20/index.html

csinva commented on tabular/src/autogluon/tabular/trainer/model_presets/presets.py:
```diff
@@ -30,6 +31,13 @@
     custom=0,
+
+    # interpretable models
+    RULEFIT=0,
+    GREEDYTREE=0,
+    RULELIST=0,
+    OPTIMALTREE=0,
+    BOOSTEDRULES=0,
```
added!

csinva commented on tabular/src/autogluon/tabular/configs/hyperparameter_configs.py:
```diff
@@ -84,6 +84,14 @@
         'AG_IMAGE_NN': {},  # TODO, Support changing the config w.r.t the preset option.
     },
+
+    interpretable={
```
added!

csinva commented on tabular/src/autogluon/tabular/models/imodels/imodels_models.py:
```diff
@@ -0,0 +1,153 @@
+from autogluon.core.utils.try_import import try_import_imodels
+
+
+class IModelsModel(AbstractModel):
```
changed!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2947,6 +2957,74 @@ def _sanitize_stack_args(self, num_bag_folds, num_bag_sets, num_stack_levels, ti
+            if summaries_filtered.shape[0] == 0:
+                summaries_filtered = summaries
+            model_name = summaries_filtered.index.values[0]  # best model is at top
+        agmodel = self._learner.load_trainer().load_model(model_name)
```
updated!

csinva commented on tabular/src/autogluon/tabular/predictor/predictor.py:
```diff
@@ -2947,6 +2957,74 @@ def _sanitize_stack_args(self, num_bag_folds, num_bag_sets, num_stack_levels, ti
+        complexities = []
+        for i in range(summaries.shape[0]):
+            model_name = summaries.index.values[i]
+            agmodel = self._learner.load_trainer().load_model(model_name)
```
updated!

jwmueller reviewed (approved):
Thanks for addressing the remaining comments. This PR looks good to me!

jwmueller commented:
@Innixma @gradientsky @yinweisu Any idea why the CI is failing to build this commit?  
The error messages seem all over the place in different commits to this PR ranging from 'core' to 'vision' to 'tabular_prediction'

jwmueller commented:
@csinva Just re-noting these thoughts here for future reference on TODOs to further improve imodels integration after this PR is merged.

The main issue I can see is the rules currently output may be nontrivial to understand for a ML novice because:

1) the format of the rules changes drastically between models, and which model is selected (ie. outputs the default rule displayed by interpretable-autoML) will be changing over different datasets, so the user would need theoretically to understand every single model's rule formatting.  Maybe for now this can easily be addressed by just having the default models under the "interpret" preset be limited to models that share a common rule format (you could still have the other models available to be trained if they are invoked in the `hyperparameters` dict). 

2) The rules output by the models are not in terms of original categorical features but rather in terms of OHE-based statements like this: `featname_level < 0.5` (eg. `workclass_5 <= 0.5`), where `level` here denotes which category-level was observed for categorical feature `featname` in this row. 
I think imodels output could be clearer if instead they were formatted like:
`featname != level` (for the case of < 0.5, or `featname == level` for the case of > 0.5).
Such a reformatting change might be best made directly in the imodels package though, instead of in autogluon, as it is a direct improvement to imodels.

3) For categorical feature, level is currently integer instead of the original string corresponding to category. This can be more easily be changed after imodels formatting has finalized a solution to issue #2 above. AutoGluon can simply pass imodels a OHE-encoded dataset where the OHE column names are: `workclass_cat1`, `workclass_cat2`,... instead of `workclass_1`, `workclass_2`,... where `cat1`, `cat2` here would be the names of the possible categories for the `workclass` column.

Innixma commented:
@csinva Can you rebase with mainline? Can't run CI due to file conflicts.

csinva commented:
Just rebased @Innixma

szha commented:
Job PR-1355-28 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/28/index.html

Innixma commented:
Thanks @csinva! I think the current plan is to aim for merging this quickly after v0.4 release so it can be part of a potential v0.4.1 release. If I find time prior to v0.4 I will test this out and see if it can make v0.4 release.

Innixma commented:
Hey @csinva, thanks for waiting. v0.4 has been released so we can move forward with this PR. Can you start by rebasing with mainline to resolve the merge conflicts?

@jwmueller Are you still of the opinion that the PR is ready to merge despite the comments you had regarding how it could be improved?

jwmueller commented:
I think there are some easy fixes that should be applied before merging during the rebase:

1. The format of the rules changes drastically between models, and which model is selected (ie. outputs the default rule displayed by interpretable-autoML) will be changing over different datasets, so the user would need theoretically to understand every single model's rule formatting. For now this can easily be addressed by just having the default models under the "interpret" preset be limited to your favorite models that share a common rule format (but you could still have the other models available to be trained if they are invoked in the hyperparameters dict).

2.  For categorical feature, `level` is currently integer-encoded version of the category name and rules are printed in terms of `level`. The easiest temporary fix for this would be whenever the user has specified "interpret" preset, autogluon's `print_interpretable_rules()` prints out a dictionary for each categorical feature that describes what the category-names are for each level of that feature, underneath the printing of the rule itself.  Furthermore, the `predictor` could have a `_category_levels_dict` attribute that contains the same dictionary (and is set during `fit()` if "interpret" preset was specified).

szha commented:
Job PR-1355-30 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1355/30/index.html

Innixma reviewed (approved):
Looks great! Thanks @csinva for all the efforts and apologies for the wait. The imodels support in AutoGluon will be available starting in v0.5 release which is planned for start of July 2022. The remaining todo items we can plan to address as a follow-up prior to v0.5 release, such as improving the readability of the rules in the case of categorical features.

Innixma commented:
Created #1748 to track the improvements to the rule printouts

csinva commented:
Thanks @Innixma! Sorry for being slow to update things....will update the other PR to make the rules more readable!
